### ISSUE TITLE
test-path-util Assertion 'endswith(p, "/lt-test-path-util")' failed

### ISSUE THREAD

**Issue by floppym** [systemd/systemd#3838]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

0fbd465f4138dd2c3c7428c23cae117a7b56568d

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

Gentoo
automake (GNU automake) 1.15
libtool (GNU libtool) 2.4.6
### Description

```
===================================
   systemd 231: ./test-suite.log
===================================

# TOTAL: 146
# PASS:  134
# SKIP:  11
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: test-path-util
====================

Assertion 'endswith(p, "/lt-test-path-util")' failed at src/test/test-path-util.c:117, function test_find_binary(). Aborting.
/bin/sh
/home/floppym/src/systemd/.libs/test-path-util
FAIL test-path-util (exit status: 134)
```

**Comment by floppym** [comment]
@fsateler Can you take a look? It looks like you updated this test in cc8b113e0ba9e8dfbc4159799133a7d32f633109.

**Comment by fsateler** [comment]
@floppym I can't reproduce. My file is named `.libs/lt-test-path-util`, and is generated when executing `./test-path-util`: `program=lt-'test-path-util'`

Can you execute the program directly?

**Comment by floppym** [comment]
Same result. On my system, `test-path-util` is a shell script that executes `.libs/test-path-util`

```
% grep program= test-path-util
  program='test-path-util'

% bash -x test-path-util
+ sed_quote_subst='s|\([`"$\\]\)|\\\1|g'
+ test -n ''
+ case `(set -o) 2>/dev/null` in
+ set -o posix
+ BIN_SH=xpg4
+ export BIN_SH
+ DUALCASE=1
+ export DUALCASE
+ unset CDPATH
+ relink_command=
+ test '' = '%%%MAGIC variable%%%'
+ test '' '!=' '%%%MAGIC variable%%%'
+ file=test-path-util
+ ECHO='printf %s\n'
+ lt_option_debug=
+ func_parse_lt_options test-path-util
+ lt_script_arg0=test-path-util
+ shift
+ test -n ''
++ /bin/sed 's%/[^/]*$%%'
++ printf '%s\n' test-path-util
+ thisdir=test-path-util
+ test xtest-path-util = xtest-path-util
+ thisdir=.
++ ls -ld test-path-util
++ /bin/sed -n 's/.*-> //p'
+ file=
+ test -n ''
+ WRAPPER_SCRIPT_BELONGS_IN_OBJDIR=no
+ test no = yes
++ cd .
++ pwd
+ absdir=/home/floppym/src/systemd
+ test -n /home/floppym/src/systemd
+ thisdir=/home/floppym/src/systemd
+ program=test-path-util
+ progdir=/home/floppym/src/systemd/.libs
+ test -f /home/floppym/src/systemd/.libs/test-path-util
+ LD_LIBRARY_PATH=/home/floppym/src/systemd/.libs:
++ /bin/sed 's/::*$//'
++ printf '%s\n' /home/floppym/src/systemd/.libs:
+ LD_LIBRARY_PATH=/home/floppym/src/systemd/.libs
+ export LD_LIBRARY_PATH
+ test '' '!=' '%%%MAGIC variable%%%'
+ func_exec_program
+ case " $* " in
+ func_exec_program_core
+ test -n ''
+ exec /home/floppym/src/systemd/.libs/test-path-util
/bin/sh
/home/floppym/src/systemd/.libs/test-path-util
Assertion 'endswith(p, "/lt-test-path-util")' failed at src/test/test-path-util.c:117, function test_find_binary(). Aborting.
zsh: abort (core dumped)  bash -x test-path-util
```

**Comment by floppym** [comment]
Also, `.libs/lt-test-path-util` does not exist here.

There seems to be some difference in the output libtool is generating between our two systems.

**Comment by fsateler** [comment]
This is strange. I supposedly have the same libtool (upstream) version: `libtool (GNU libtool) 2.4.6 Debian-2.4.6-1`

I don't see anyhing in the debian patches that would alter this behavior. Maybe gentoo does?

**Comment by floppym** [comment]
I see a handful of backported patches from the libtool master branch, but nothing that looks obviously related to this issue.

https://gitweb.gentoo.org/repo/gentoo.git/tree/sys-devel/libtool/files

**Comment by floppym** [comment]
libtool has some concept called "fast installation", which seems to control what the binary in `.libs` gets called.

From libtool:

```
        if test yes = "$fast_install"; then
          $ECHO "\
  program=lt-'$outputname'$exeext
  progdir=\"\$thisdir/$objdir\"
...
        else
          $ECHO "\
  program='$outputname'
  progdir=\"\$thisdir/$objdir\"
"
        fi
```

I'm not sure why this `fast_install` variable gets set differently on my system. However, it seems bad to assume that the binary will always be prefixed with `lt-` given that it seems to be configurable somehow.

**Comment by floppym** [comment]
I think I figured out why Gentoo has a different default from Debian here. I think we patch binutils to generate `DT_RUNPATH` instead of `DT_RPATH` when passed the `-rpath` option.

This means libtool does not need to prefix the binary names to allow `LD_LIBRARY_PATH` to override the runpath on the binary. You can see the behavior change by searching for `shlibpath_overrides_runpath` in libtool.m4.

In summary: depending on the Linux distro, libtool may generate binaries with or without the `lt-` prefix. This test must be changed to accommodate that.
